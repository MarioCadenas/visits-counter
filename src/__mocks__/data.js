export const mockData = {
  loadingTrue: {
    loading: true,
  },
  loadingFalse: {
    loading: false,
  },
  emptyVisits: {
    visits: []
  },
  notEmptyVisits: {
    visits: [{ id: '123', count: 5 }]
  }
};
export const mockFunctions = {
  createVisit: jest.fn(() => ({ data: { createVisit: { count: 1 } } })),
  updateVisit: jest.fn(() => ({ data: { updateVisit: { count: mockData.notEmptyVisits.visits[0].count + 1 } } })),
};
