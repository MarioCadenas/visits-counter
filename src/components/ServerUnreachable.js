import React from 'react';
import logo from '../logo.svg';

export const ServerUnreachable = () => (
  <div className="App">
    <header className="App-header">
      <img src={logo} className="App-logo" alt="logo" />
      <h1>El servidor no responde</h1>
    </header>
  </div>
);
