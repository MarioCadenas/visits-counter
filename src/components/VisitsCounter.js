import React, { Component } from 'react';

export default class VisitsCounter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      id: props.visit ? props.visit.id : null,
      count: props.visit ? props.visit.count : 0
    }
  }

  async componentDidMount() {
    const { props: { updateVisit, createVisit }, state: { id, count: currentCount } } = this;

    if (!id) {
      const { data: { createVisit: { count } } } = await createVisit({ variables: { count: currentCount + 1 } });
      return this.setState({ count });
    }

    const { data: { updateVisit: { count } } } = await updateVisit({ variables: { id } });
    return this.setState({ count });
  }

  render() {
    const { state: { count } } = this;
    return <h1>{`Visitas totales: ${count}`}</h1>
  }
}
