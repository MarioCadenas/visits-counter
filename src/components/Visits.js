import React, { Component } from 'react';
import { gql } from 'apollo-boost';
import { graphql, compose } from 'react-apollo';
import { Loader } from './Loader';
import { getVisitsQuery } from '../graphql/queries';
import { updateVisitMutation, createVisitMutation } from '../graphql/mutations';
import VisitsCounter from './VisitsCounter';

export class VisitsPage extends Component {
  render() {
    if (this.props.getVisits.loading) {
      return <Loader />;
    }

    const { props: { getVisits: { visits: [visit] } } } = this;

    return (
      <VisitsCounter
        visit={visit}
        createVisit={this.props.createVisit}
        updateVisit={this.props.updateVisit}
      />
    )
  }
}

export default compose(
  graphql(getVisitsQuery(gql), { name: 'getVisits' }),
  graphql(createVisitMutation(gql), { name: 'createVisit' }),
  graphql(updateVisitMutation(gql), { name: 'updateVisit'})
)(VisitsPage);
