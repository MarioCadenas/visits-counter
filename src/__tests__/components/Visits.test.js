import React from 'react';
import { shallow, mount } from 'enzyme';
import { VisitsPage } from '../../components/Visits';
import { mockData, mockFunctions } from '../../__mocks__/data';

describe('Visits Component', () => {
  it('should correctly render visit component when loading', () => {
    const { loadingTrue, emptyVisits } = mockData;
    const { createVisit, updateVisit } = mockFunctions;
    const wrapper = shallow(
      <VisitsPage
        getVisits={{ ...loadingTrue, ...emptyVisits }}
        createVisit={createVisit}
        updateVisit={updateVisit}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should correctly render visit component when not loading', () => {
    const { loadingFalse, notEmptyVisits } = mockData;
    const { createVisit, updateVisit } = mockFunctions;
    const wrapper = shallow(
      <VisitsPage
        getVisits={{ ...loadingFalse, ...notEmptyVisits }}
        createVisit={createVisit}
        updateVisit={updateVisit}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.prop('createVisit')).toEqual(createVisit);
    expect(wrapper.prop('updateVisit')).toEqual(updateVisit);
  });
  it('should call createVisit when empty visits', () => {
    const { loadingFalse, emptyVisits } = mockData;
    const { createVisit, updateVisit } = mockFunctions;
    mount(
      <VisitsPage
        getVisits={{ ...loadingFalse, ...emptyVisits }}
        createVisit={createVisit}
        updateVisit={updateVisit}
      />
    );
    expect(createVisit).toHaveBeenCalledWith({ variables: { count: 1 } });
  });
  it('should call updateVisit when visits counter exists', () => {
    const { loadingFalse, notEmptyVisits } = mockData;
    const { createVisit, updateVisit } = mockFunctions;
    mount(
      <VisitsPage
        getVisits={{ ...loadingFalse, ...notEmptyVisits }}
        createVisit={createVisit}
        updateVisit={updateVisit}
      />
    );
    expect(updateVisit).toHaveBeenCalledWith({ variables: { id: notEmptyVisits.visits[0].id } });
  });
});
