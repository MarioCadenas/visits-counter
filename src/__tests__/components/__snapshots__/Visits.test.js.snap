// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visits Component should correctly render visit component when loading 1`] = `<Loader />`;

exports[`Visits Component should correctly render visit component when not loading 1`] = `
<VisitsCounter
  createVisit={[MockFunction]}
  updateVisit={[MockFunction]}
  visit={
    Object {
      "count": 5,
      "id": "123",
    }
  }
/>
`;
